src/relations/db.py

```diff
@@ -0,0 +1,414 @@
+        self.admin = admin
+
+    def _on_relation_joined(self, join_event: RelationJoinedEvent):
+        """Handle db-relation-joined event."""
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm.backend_postgres:
+            # We can't relate an app to the backend database without a backend postgres relation
+            logger.warning("waiting for backend-database relation to connect")
+            join_event.defer()
+            return
+
+        logger.info(f"Setting up {self.relation_name} relation")
+        logger.warning(
+            f"DEPRECATION WARNING - {self.relation_name} is a legacy relation, and will be deprecated in a future release. "
+        )
+
+        cfg = self.charm.read_pgb_config()
+
+        relation_data = join_event.relation.data
+        pgb_unit_databag = relation_data[self.charm.unit]
+        pgb_app_databag = relation_data[self.charm.app]
+        remote_app_databag = relation_data[join_event.app]
+
+        # Do not allow apps requesting extensions to be installed.
+        if "extensions" in remote_app_databag:
+            logger.error(
+                "ERROR - `extensions` cannot be requested through relations"
+                " - they should be installed through a database charm config in the future"
+            )
+            self.charm.unit.status = BlockedStatus(
+                "bad relation request - remote app requested extensions, which are unsupported."
+            )
+            return
+
+        database = remote_app_databag.get("database")
+        if database is None:
+            logger.warning("No database name provided in app databag")
+            join_event.defer()
+            return
+
+        user = self.generate_username(join_event)
+        password = pgb.generate_password()
+
+        # Create user in pgbouncer config
+        self.charm.add_user(
+            user,
+            password=password,
+            admin=self.admin,
+            cfg=cfg,
+            render_cfg=True,
+            reload_pgbouncer=True,
+        )
+
+        # Create user and database in backend postgresql database
+        try:
+            init_msg = f"initialising database and user for {self.relation_name} relation"
+            self.charm.unit.status = MaintenanceStatus(init_msg)
+            logger.info(init_msg)
+
+            self.charm.backend_postgres.create_user(user, password, admin=self.admin)
+            self.charm.backend_postgres.create_database(database, user)
+
+            created_msg = f"database and user for {self.relation_name} relation created"
+            self.charm.unit.status = ActiveStatus(created_msg)
+            logger.info(created_msg)
+        except (PostgreSQLCreateDatabaseError, PostgreSQLCreateUserError):
+            err_msg = f"failed to create database or user for {self.relation_name}"
+            logger.error(err_msg)
+            self.charm.unit.status = BlockedStatus(err_msg)
+            join_event.defer()
```

Comment:
Maybe it is better to not defer the event? I am thinking about some error that is not transient.

Reply:
You're right - this was originally written as a catch if backend_postgres wasn't finished implementing but we don't start the hook unless it's ready to go, so I'll remove the deferral.